File: tools/meal_browser.py
"""Render one browsable card per restaurant order, with a no-JavaScript fallback."""
from html import escape

FILTERS = [
    ('', 'All meals', ''),
    ('protein', 'High protein', 'highest-protein-fast-food'),
    ('light', '400 calories or less', 'fast-food-under-400-calories'),
    ('energy', '600+ calories', 'high-calorie-fast-food'),
    ('vegetarian', 'Vegetarian', 'vegetarian-fast-food'),
    ('plant', 'Plant-based', 'plant-based-fast-food'),
    ('breakfast', 'Breakfast', 'healthy-fast-food-breakfast'),
    ('fibre', 'High fiber', 'highest-fibre-fast-food'),
    ('lowsodium', 'Lower sodium', 'lowest-sodium-fast-food'),
]


def render(meals):
    out = ['<!--MEALS:START-->', '<div class="container meal-browser-shell">',
           '<details class="meal-database-disclosure" id="browse-meals">',
           f'<summary><strong>Browse {len(meals)} meals</strong></summary>',
           '<div class="meal-browser" data-meal-browser>',
           '<div class="meal-browser-controls" hidden>',
           '<label for="browse-query">Search meals or restaurants<input type="search" id="browse-query" placeholder="e.g. chicken" autocomplete="off"></label>',
           '<label for="browse-filter">Show<select id="browse-filter">']
    out.extend(f'<option value="{value}">{label}</option>' for value, label, _ in FILTERS)
    out.extend(['</select></label></div>',
                '<p class="browse-status" role="status" aria-live="polite" aria-atomic="true" hidden></p>',
                '<div class="browse-meal-grid">'])
    chain_order = ['Chipotle', 'Chick-fil-A', 'Subway', 'Panera', 'Panda Express']
    def order_key(meal):
        complete = all(meal[k] is not None for k in ('cal', 'p', 'f', 'na'))
        rank = chain_order.index(meal['chain']) if meal['chain'] in chain_order else len(chain_order)
        return (not complete, rank, meal['chain'], meal['name'])
    for meal in sorted(meals, key=order_key):
        # Editorial prefixes are not part of the restaurant's order name.
        name = meal['name'].replace('High-protein bulking order: ', '')
        tags = meal['t'] + meal['diet'] + ([meal['meal']] if meal['meal'] == 'breakfast' else [])
        slug = meal['url'].split('-healthy')[0]
        out.append(f'<article class="browse-meal" data-tags="{escape(" ".join(tags))}" '
                   f'data-search="{escape(meal["chain"] + " " + name, quote=True)}" '
                   f'data-protein="{meal["p"] if meal["p"] is not None else -1}" '
                   f'data-calories="{meal["cal"] if meal["cal"] is not None else -1}" '
                   f'data-fiber="{meal["f"] if meal["f"] is not None else -1}" '
                   f'data-sodium="{meal["na"] if meal["na"] is not None else 99999}">')
        out.append(f'<div class="browse-meal-brand"><img src="images/restaurant-logos/{escape(slug)}.png" '
                   f'width="36" height="36" alt="" loading="lazy"><span>{escape(meal["chain"])}</span></div>')
        out.append(f'<h3>{escape(name)}</h3><dl class="browse-meal-stats">')
        for key, label, unit in [('cal', 'Calories', ''), ('p', 'Protein', ' g'), ('f', 'Fiber', ' g'), ('na', 'Sodium', ' mg')]:
            value = 'Not listed' if meal[key] is None else f'{meal[key]:,}{unit}'
            out.append(f'<div><dt>{label}</dt><dd>{value}</dd></div>')
        out.append(f'</dl><a class="btn action-link" href="{escape(meal["url"])}">View menu</a></article>')
    out.extend(['</div>', '<p class="browse-empty" hidden>No meals found. Try another search or filter.</p>',
                '<button class="btn browse-more" type="button" hidden>Show more meals</button>',
                '<details class="browse-nutrition-note"><summary>Nutrition &amp; filters</summary>',
                '<p>Numbers are for the listed order; portions and substitutions can change them. Check allergens with the restaurant.</p>',
                '<p>High protein: 25 g or more. High fiber: 5 g or more. Lower sodium: up to 600 mg. Lower-calorie and lower-sodium filters include meals with at least 250 calories and 15 g protein.</p>',
                '<a href="sources.html">Nutrition sources</a></details>',
                '</div></details></div>', '<script src="js/meal-browser.js" defer></script>', '<!--MEALS:END-->'])
    return '\n'.join(out)

File: js/meal-browser.js
/* Progressive enhancement: one list, simple filters, six meals at a time. */
(function () {
  'use strict';
  var root = document.querySelector('[data-meal-browser]');
  if (!root) return;
  var grid = root.querySelector('.browse-meal-grid');
  var cards = Array.from(grid.children);
  var query = root.querySelector('#browse-query');
  var filter = root.querySelector('#browse-filter');
  var more = root.querySelector('.browse-more');
  var status = root.querySelector('.browse-status');
  var limit = 6;
  function revealAfterAction(element) {
    var header = document.querySelector('.site-header');
    var clearance = header ? header.getBoundingClientRect().height + 24 : 24;
    window.scrollTo({top: Math.max(0, window.scrollY + element.getBoundingClientRect().top - clearance), behavior: 'auto'});
  }
  function normalize(text) { return text.toLowerCase().replace(/[’']/g, '').replace(/[^a-z0-9]+/g, ' ').trim(); }
  function update(append) {
    var terms = normalize(query.value).split(' ').filter(Boolean);
    var tag = filter.value;
    var matched = cards.filter(function (card) {
      return (!tag || card.dataset.tags.split(' ').indexOf(tag) !== -1) &&
        terms.every(function (word) { return normalize(card.dataset.search).indexOf(word) !== -1; });
    });
    var sort = {protein: ['protein', -1], light: ['calories', 1], energy: ['calories', -1], fibre: ['fiber', -1], lowsodium: ['sodium', 1]}[tag];
    if (sort) matched.sort(function (a, b) { return (Number(a.dataset[sort[0]]) - Number(b.dataset[sort[0]])) * sort[1]; });
    var firstNew = matched[limit - 6];
    cards.forEach(function (card) { card.hidden = true; });
    matched.forEach(function (card, i) { grid.appendChild(card); card.hidden = i >= limit; });
    status.textContent = matched.length ? Math.min(limit, matched.length) + ' of ' + matched.length + ' meals' : 'No matching meals';
    more.hidden = matched.length <= limit;
    root.querySelector('.browse-empty').hidden = !!matched.length;
    if (append && firstNew) {
      var heading = firstNew.querySelector('h3');
      heading.tabIndex = -1;
      heading.focus({preventScroll: true});
      revealAfterAction(firstNew);
    }
  }
  function reset() { limit = 6; update(false); }
  query.addEventListener('input', reset);
  filter.addEventListener('change', reset);
  more.addEventListener('click', function () { limit += 6; update(true); });
  root.querySelector('.meal-browser-controls').hidden = false;
  status.hidden = false;
  // Preserve older links to a ranked list as a filter on the new browser.
  var oldLists = {'highest-protein-fast-food':'protein','fast-food-under-400-calories':'light','high-calorie-fast-food':'energy','vegetarian-fast-food':'vegetarian','plant-based-fast-food':'plant','healthy-fast-food-breakfast':'breakfast','highest-fibre-fast-food':'fibre','lowest-sodium-fast-food':'lowsodium'};
  function restoreHash() {
    var tag = oldLists[location.hash.slice(1)];
    if (tag) {
      filter.value = tag;
      root.closest('details').open = true;
      reset();
      revealAfterAction(root.closest('details'));
    }
  }
  reset();
  restoreHash();
  window.addEventListener('hashchange', restoreHash);
}());
